### PR TITLE
Use raw public key instead of compressed

### PR DESCRIPTION
**Pull request by fubuloubu**
Fixes: #4216

### CODE REVIEW

File: primitives/core/src/ecdsa.rs [hunk 1]
```diff
@@ -332,16 +332,27 @@ pub enum PublicError {
...
+	pub fn from_raw(data: [u8; 64]) -> Self {
 		Public(data)
 	}
 
+	/// A new instance from the given 65-byte `data`.
+	///
+	/// NOTE: No checking goes on to ensure this is a real public key. Only use it if
+	/// you are certain that the array actually is a pubkey. GIGO!
+    pub fn from_full(data: [u8; 65]) -> Self {
+        let raw_key = &data[1..];
```
Comment (bkchr) [inline primitives/core/src/ecdsa.rs]: Whitespaces

**Comment by fubuloubu** [comment]
Also, as noted in https://github.com/paritytech/substrate/issues/4216#issuecomment-559147275 using `keccak256` instead of `blake2_256` in `AccountId` hashing logic would make this _more_ Ethereum-compatible, but as I noted in my response it would be better if `AccountId` wasn't assumed 32 bytes (instead just `Sized`) and was user configurable with how addresses are encoded.

I could add `keccak_256` to this PR though.

**Comment by gavofyork** [comment]
`AccountId` is:

```
type AccountId: Parameter + Member + MaybeSerializeDeserialize + Debug + MaybeDisplay + Ord + Default;
```

So basically, you can have it be any type you want. There's no need for this PR, though since it doesn't really break anything and makes things strictly more compatible I'm ok merging it. But you can in any case make a 20-byte `AccountId` if you really want and you can make a new `Signature` type which is compatible with it (and a new `MultiSignature` if you want to support other crypto at the same time, too).

**Comment by gavofyork** [comment]
For now, though, some of the CI is failing and will need to be fixed before it can be merged.

**Comment by fubuloubu** [comment]
> [Y]ou can in any case make a 20-byte `AccountId` and use sha256, keccak256, ripemd-160 or whatever else if you really want and you can make a new `Signature` type which is compatible with it (and a new `MultiSignature` if you want to support other crypto at the same time, too).

Ah, well I will have to try a different method of doing what I want. I was using `MultiSigner` and `MultiSignature`, which builds in that presumption of the `blake2` hash, and I assumed there was no other way to do it, but I see where I am wrong though. This PR is definitely still foundational however.

Will fix CI issues.

File: primitives/core/src/ecdsa.rs [hunk 2]
```diff
@@ -558,7 +569,7 @@ mod test {
 		);
 		let public = pair.public();
 		assert_eq!(public, Public::from_raw(
-			hex!("028db55b05db86c0b1786ca49f095d76344c9e6056b2f02701a7e7f3c20aabfd91")
+			hex!("8db55b05db86c0b1786ca49f095d76344c9e6056b2f02701a7e7f3c20aabfd913ebbe148dd17c56551a52952371071a6c604b3f3abe8f2c8fa742158ea6dd7d4")
 		));
 		let message = b"";
 		let signature = hex!("3dde91174bd9359027be59a428b8146513df80a2a3c7eda2194f64de04a69ab97b753169e94db6ffd50921a2668a48b94ca11e3d32c1ff19cfe88890aa7e8f3c00");
```
Comment (fubuloubu) [inline primitives/core/src/ecdsa.rs]: One of the CI failures is due to line width of the line 3 above this one. How did this one get configured not to trip the line size check?

**Comment by fubuloubu** [comment]
As noted above, the 120 char line limit is failing because I made a hex string larger. It doesn't seem like I should be splitting it up, so it is probably best to ignore this error?

The other two errors do not seem to be my fault, as far as I can tell.

Please advise.

File: primitives/core/src/ecdsa.rs (hunk 2 above)
Comment (gavofyork) [reply primitives/core/src/ecdsa.rs]: yeah unfortunately i don't think it's possible, so can ignore it.

**Pull request merged**